### ISSUE TITLE
Why pad the input in conv() ? and why different D-layers have different strides? 

### ISSUE THREAD

**Issue by qixiang109** [affinelayer/pix2pix-tensorflow#84]
in the conv() function, the input is padded with a zero contour:

`padded_input = tf.pad(batch_input, [[0, 0], [1, 1], [1, 1], [0, 0]], mode="CONSTANT")
conv = tf.nn.conv2d(padded_input, filter, [1, stride, stride, 1], padding="VALID")
`

and in the create_discriminator() function, different layers have different strides:

`
for i in range(n_layers):
            with tf.variable_scope("layer_%d" % (len(layers) + 1)):
                out_channels = a.ndf * min(2**(i+1), 8)
                stride = 1 if i == n_layers - 1 else 2  # last layer here has stride 1
...
`

I really don't know why we need these two settings? They together make the tensor shapes of the Dnet changes like 256-128-64-32-31-30....

**Comment by qixiang109** [comment]
I think  I have relized the effect of the [[0, 0], [1, 1], [1, 1], [0, 0]] padding: a spatial convolution layer's outputs are tensors of size (input_size - kernel_size)/stride+1. When stride=2 and kernel_size=4, such a padding makes the output size equals (input_size +2 -4)/2+1, which is precisely 1/2 input_size. This means no pixel is dropped during convolution.
But I've still no idea why the strides changes in the D net .

**Comment by julien2512** [comment]
That link helps me to understand : https://stackoverflow.com/questions/34642595/tensorflow-strides-argument

Use the stride=1 is a choice to make 32=>31=>30 (as the layer_5 does it too).

With a stride=2, it would be 32=>16=>8 (if the layer_5 does it too).

**Comment by qixiang109** [comment]
@julien2512 yes I have understood it by my self, thank you anyway!
